operators/src/adapters/raster_overlap_adapter.rs

```diff
@@ -132,36 +149,47 @@ where
         let mut this = self.project();
 
         if this.running_future.as_ref().is_none() {
-            if *this.current_spatial_tile >= this.spatial_tiles.len() {
+            // there is no future running / stream processing
+            if *this.current_spatial_tile >= this.tiles_to_produce.len() {
+                // we iterated through all the tiles, now we need to move time forward
                 *this.current_spatial_tile = 0;
-                *this.time_start = *this.time_end;
-                *this.time_end = None;
+
+                // make time progress
+                let (n_t_start, n_t_end) = match (*this.time_start, *this.time_end) {
+                    (Some(t_start), Some(t_end)) if t_start == t_end => (Some(t_start + 1), None),
+                    (Some(t_start), Some(t_end)) if t_start < t_end => (Some(t_end), None),
+                    (_, _) => (None, None),
+                };
+
+                *this.time_start = n_t_start;
+                *this.time_end = n_t_end;
             }
 
             if let Some(t_start) = *this.time_start {
                 if t_start >= this.query_rect.time_interval.end() {
                     *this.ended = true;
                 } else {
-                    // TODO: maybe this whole block should also be a future!
-                    let current_spatial_tile_info = this.spatial_tiles[*this.current_spatial_tile];
+                    let current_spatial_tile_info =
+                        this.tiles_to_produce[*this.current_spatial_tile];
 
                     let fold_tile_spec = TileInformation {
                         tile_size_in_pixels: current_spatial_tile_info.tile_size_in_pixels,
                         global_tile_position: current_spatial_tile_info.global_tile_position,
                         global_geo_transform: current_spatial_tile_info.global_geo_transform,
                     };
 
-                    let tqr = (this.tile_query_rewrite)(
-                        &current_spatial_tile_info,
-                        &this.query_rect,
+                    let tqr = this.sub_query.tile_query_rectangle(
+                        fold_tile_spec,
+                        *this.query_rect,
                         t_start,
                     )?;
 
-                    let qs = this.source.raster_query(tqr, *this.query_ctx)?;
+                    // TODO: this schould also be a future. We can chain it with the query if we find a way to store it in the running future
+                    let accu = this.sub_query.new_fold_accu(fold_tile_spec, tqr)?;
 
-                    let accu = (this.accu_creator)(fold_tile_spec, tqr);
+                    let qs = this.source.raster_query(tqr, *this.query_ctx)?;
 
-                    let ttf = qs.fold(accu, this.fold_fn.clone());
+                    let ttf = qs.try_fold(accu, this.sub_query.fold_method());
 
                     this.running_future.set(Some(ttf));
```

Comment:
You often use only super short variable names here, like `tqr`, `qs`, `ttf`, `r` and `rv`. I guess in two weeks from now, this is even hard to understand for you :wink:.

Reply:
0b50d56